cityofaustin/atd-moped
# Update `moped_project` and `moped_proj_components` on child feature edits

## Associated issues

This PR hopes to close https://github.com/cityofaustin/atd-data-tech/issues/14617. 

## Testing

Testing can be completed locally.

* While testing, keep in mind that these triggers are set to run on INSERT and UPDATE to the underlying feature tables. An INSERT into one of those tables is an "update"-like operation to the parent project and component records, which is why we won't be editing `created_at` and `created_by_user_id` fields on the project or component records.

* Spin up the local database and make sure migrations are applied cleanly.
* Add a feature to a project, noting the project number.

* You can use your SQL client to check the pertinent fields in the projects and components table by plugging the project number into this query:

```sql
select 
  projects.project_id, components.component_id,
  projects.updated_at as project_updated_at, projects.updated_by_user_id as project_updated_by,
  components.updated_at as component_updated_at, components.updated_by_user_id as component_updated_by
from moped_project projects
left join moped_proj_components components on (components.project_id = projects.project_id)
where projects.project_id = 119
```

* Run this query to make sure the updated fields are updated in the tables. The "project_" fields are in the projects table, so if the project you're editing has multiple components, you'll see the `project_updated_at` and `project_updated_by` value repeated for each component. ✅
* Add new features (components) and update the geometry of existing components, and watch how these updated_by fields are updated by these triggers and function.

## Utility

I made a quick `reload.sh` that was invaluable in constructing this PR. We may want to enhance the `hasura-cluster` script with something along these lines at some point.

```
#!/bin/bash

./hasura-cluster stop
./hasura-cluster start_only
docker stop moped-database-hasura-1
echo 'DROP DATABASE moped; CREATE DATABASE moped' | docker exec -i moped-database-moped-pgsql-1 psql -U moped postgres;
docker exec -i moped-database-moped-pgsql-1 psql -U moped moped < production.sql; 
docker start moped-database-hasura-1
sleep 5
hasura migrate apply
hasura metadata apply
#echo 'UPDATE moped_users set user_id = 1 where user_id = 60;' | docker exec -i moped-database-moped-pgsql-1 psql -U moped moped;
```

---
#### Ship list
